update ds/dsu (#44)
OI-wiki/i18n: docs/ds/dsu.md

author: HeRaNO, JuicyMio, Xeonacid, sailordiary, ouuan

Disjoint set is a tree-shaped data structure. As the name suggests, it is used to deal with some disjoint **merge** and **query** problems.
It supports two operations:

-   Find: determine which subset an element is in.

-   Union: merge two subsets into one set.

!!! warning

>     The disjoint set does not support the set separation. However, it supports the deletion of a single element in the set after modification (see [UVA11987 Almost Union-Find](https://onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&page=show_problem&problem=3138) for details). Dynamically opening nodes on segment tee can also realize persistent disjoint set.

Reviewer: open 🌚

## Initialization

```cpp

> void makeSet(inßt size) {

Reviewer: 🌚 这个 typo 是怎么打出来的

Author: ...MacBook的键盘🌚 🌚 最近刚换mbp不太习惯

  for (int i = 0; i < size; i++) fa[i] = i;  // i is in its own set
  return;
}
```

## Find

> First let's hear a story: Several families hold banquets and there are too many people because they usually live a long life. Due to long time separation and aging, these people gradually forget their relatives and only remember who their father is. The father of the oldest person (called "ancestor") has passed away, and he only knows that he is an ancestor. . In order to determine which family they belonged to, they came up with a way, as long as they asked their father if they were ancestors, and asked them generation by generation until they asked their ancestors. If you want to know whether two people are in the same family, you only need to see whether the ancestors of the two are the same.

Reviewer: ```suggestion
First let's hear a story: Several families hold banquets and there are too many people because they usually live a long life. Due to long time separation and aging, these people gradually forget their relatives and only remember who their father is. The father of the oldest person (called "ancestor") has passed away, and he only knows that he is an ancestor. In order to determine which family they belonged to, they came up with a way, as long as they asked their father if they were ancestors, and asked them generation by generation until they asked their ancestors. If you want to know whether two people are in the same family, you only need to see whether the ancestors of the two are the same.
```
讲故事可以用过去时吧 🤔 
diff 是一个多余的句号

> In this way of thinking, the find algorithm of disjoint algorithm was born.

Reviewer: ```suggestion
In this way of thinking, the find algorithm of disjoint set was born.
```
🤔

Here we offer a C++ implementation for your reference:

```cpp
int fa[MAXN];  // record who is who's father, with special rules that the father of an ancestor is himself
int find(int x) {
  // find the ancestors of x
  if (fa[x] == x)  // return if x is an ancestor
    return x;
  else
    return find(fa[x]);  // if not, then x’s dad asks x’s grandpa
}
```

Obviously this will eventually return the ancestor of $x$ .

### Path compression

这样的确可以达成目的，但是显然效率实在太低。为什么呢？因为我们使用了太多没用的信息，我的祖先是谁与我父亲是谁没什么关系，这样一层一层找太浪费时间，不如我直接当祖先的儿子，问一次就可以出结果了。甚至祖先是谁都无所谓，只要这个人可以代表我们家族就能得到想要的效果。 **把在路径上的每个节点都直接连接到根上** ，这就是路径压缩。

此处给出一种 C++ 的参考实现：

> 

Reviewer: 中文没删

> 

Reviewer: 中文没删

The above way can indeed achieve the goal, but obviously the efficiency is too low. Why? Because we use too much useless information. Who my ancestors are and who my father is has nothing to do with it, and it is a waste of time to find generaion by generation. It’s better to let "me" be the "son" of my ancestors and obtain results. It doesn't even matter who the ancestor is, as long as this person can represent the family. **Connecting each node on the path directly to the root**, and this is path compression.

Here we offer a C++ implementation for your reference:

```cpp
int find(int x) {
  if (x != fa[x])  // x is not its own father, that is, x is not a representative of the set

>     fa[x] = find(fa[x]);  // find the ancestors of x until you find a representative, so the path is compressed

Reviewer: 顺手用 so 不太合适吧
and compress the path along the way

  return fa[x];
}
```

See these two figures:

![p1](./images/dsu1.png)

![p2](./images/dsu2-en.png)

## Union

> At the banquet, the ancestor of one family suddenly said to another family: Our two families have such a good relationship, so what about forming one family? The other family also gladly accepted.

Reviewer: merging into?

> As we said before, we don’t care who the ancestors are, so as long as one of the ancestors becomes the son of the other ancestor.

Reviewer: as long as 后面要接子句的吧
So it's only need to

Author: 把so 直接删掉？用so it only needs to 感觉怪怪的 :thinking:

Reviewer: 但是倒得有个连词，不然得拆成两个句子了吧

Reviewer: so it only needs to make an ancestor to be the **child** of another one.

Here we offer a C++ implementation for your reference:

```cpp
void unionSet(int x, int y) {
  // merge the families of x and y
  x = find(x);
  y = find(y);

>   if (x == y)  // do not care if it was in the same family

Reviewer: do nothing

    return;
  fa[x] = y;  // turn x's ancestor into y's ancestor's son
}
```

### Heuristic merge (merging by rank)

> One of the ancestors suddenly had an interesting idea:"Your family has fewer people. It is more convenient to move to our family. It would be too much trouble if we move there."

Reviewer: ```suggestion
One of the ancestors suddenly had an interesting idea: "Your family has fewer people. It is more convenient to move to our family because it would be too troublesome if we move there."
```

If only heuristic merging is used instead of path compression, the time complexity is $\Theta(m\log n)$ . Since a single merge of path compression may cause a large number of modifications, sometimes path compression is not suitable for use. For example, in the persistent disjoint set, segment tree divide and conquer + disjoint set, only heuristic merge is generally used.

Here is a C++ reference implementation, which uses the number of nodes as the evaluation function:

```cpp
std::vector<int> size(N, 1);  // record and initialize the size of the subtree to 1
void unionSet(int x, int y) {
  int xx = find(x), yy = find(y);
  if (xx == yy) return;
  if (size[xx] > size[yy])  // ensure that the small ones merge into the big ones
    swap(xx, yy);
  fa[xx] = yy;
  size[yy] += size[xx];
}
```

## Time complexity and space complexity

### Time complexity 

After using path compression and heuristic merging at the same time, the average time of each operation is only $O(\alpha(n))$ , where $\alpha$ is the inverse function of Ackerman function, and its growth is extremely slow. That is to say, the average running time of a single operation can be considered as a small constant.

>  [Ackermann function](https://en.wikipedia.org/wiki/Ackermann_function)  $A(n, m)$ is defined as:

Reviewer: 这行合到上一段的正文里吧，然后变成 [Ackermann function](https://en.wikipedia.org/wiki/Ackermann_function), which is defined as the formula below and its growth is ....

>  [Ackermann function](https://en.wikipedia.org/wiki/Ackermann_function)  $A(n, m)$ is defined as:

Reviewer: ```suggestion
The [Ackermann function](https://en.wikipedia.org/wiki/Ackermann_function), $A(n, m)$, is defined as:
```
